// distube/dist/util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isRecord = exports.isObject = exports.entersState = exports.checkInvalidKey = exports.isClientInstance = exports.resolveGuildID = exports.resolveGuildId = exports.isGuildInstance = exports.isSupportedVoiceChannel = exports.isMessageInstance = exports.isTextChannelInstance = exports.isMemberInstance = exports.isSnowflake = exports.isVoiceChannelEmpty = exports.checkIntents = exports.isURL = exports.parseNumber = exports.toSecond = exports.formatDuration = void 0;
const url_1 = require("url");
const _1 = require(".");
const discord_js_1 = require("discord.js");
const formatInt = (int) => (int < 10 ? `0${int}` : int);
/**
 * Format duration to string
 * @param {number} sec Duration in seconds
 * @returns {string}
 */
function formatDuration(sec) {
    if (!sec || !Number(sec))
        return "00:00";
    const seconds = Math.round(sec % 60);
    const minutes = Math.floor((sec % 3600) / 60);
    const hours = Math.floor(sec / 3600);
    if (hours > 0)
        return `${formatInt(hours)}:${formatInt(minutes)}:${formatInt(seconds)}`;
    if (minutes > 0)
        return `${formatInt(minutes)}:${formatInt(seconds)}`;
    return `00:${formatInt(seconds)}`;
}
exports.formatDuration = formatDuration;
/**
 * Convert formatted duration to seconds
 * @param {*} input Formatted duration string
 * @returns {number}
 */
function toSecond(input) {
    if (!input)
        return 0;
    if (typeof input !== "string")
        return Number(input) || 0;
    if (input.match(/:/g)) {
        const time = input.split(":").reverse();
        let s = 0;
        for (let i = 0; i < 3; i++)
            if (time[i])
                s += Number(time[i].replace(/[^\d.]+/g, "")) * Math.pow(60, i);
        if (time.length > 3)
            s += Number(time[3].replace(/[^\d.]+/g, "")) * 24 * 60 * 60;
        return s;
    }
    else {
        return Number(input.replace(/[^\d.]+/g, "")) || 0;
    }
}
exports.toSecond = toSecond;
/**
 * Parse number from input
 * @param {*} input Any
 * @returns {number}
 */
function parseNumber(input) {
    if (typeof input === "string")
        return Number(input.replace(/[^\d.]+/g, "")) || 0;
    return Number(input) || 0;
}
exports.parseNumber = parseNumber;
/**
 * Check if the string is an URL
 * @param {string} input input
 * @returns {boolean}
 */
function isURL(input) {
    if (typeof input !== "string" || input.includes(" "))
        return false;
    try {
        const url = new url_1.URL(input);
        if (!["https:", "http:"].includes(url.protocol) || !url.host)
            return false;
    }
    catch {
        return false;
    }
    return true;
}
exports.isURL = isURL;
/**
 * Check if the Client has enough intents to using DisTube
 * @param {ClientOptions} options options
 */
function checkIntents(options) {
    const requiredIntents = ["GUILD_VOICE_STATES"];
    const bitfield = options.intents ?? options?.ws?.intents;
    if (typeof bitfield === "undefined")
        return;
    const intents = new discord_js_1.Intents(bitfield);
    for (const intent of requiredIntents) {
        if (!intents.has(intent))
            throw new _1.DisTubeError("MISSING_INTENTS", intent.toString());
    }
}
exports.checkIntents = checkIntents;
/**
 * Check if the voice channel is empty
 * @param {Discord.VoiceState} voiceState voiceState
 * @returns {boolean}
 */
function isVoiceChannelEmpty(voiceState) {
    const voiceChannel = voiceState.guild?.me?.voice?.channel;
    if (!voiceChannel)
        return false;
    const members = voiceChannel.members.filter(m => !m.user.bot);
    return !members.size;
}
exports.isVoiceChannelEmpty = isVoiceChannelEmpty;
function isSnowflake(id) {
    try {
        return discord_js_1.SnowflakeUtil.deconstruct(id).timestamp > discord_js_1.SnowflakeUtil.EPOCH;
    }
    catch {
        return false;
    }
}
exports.isSnowflake = isSnowflake;
function isMemberInstance(member) {
    return (!!member &&
        isSnowflake(member.id) &&
        isSnowflake(member.guild?.id) &&
        isSnowflake(member.user?.id) &&
        member.id === member.user.id);
}
exports.isMemberInstance = isMemberInstance;
function isTextChannelInstance(channel) {
    return (!!channel &&
        isSnowflake(channel.id) &&
        isSnowflake(channel.guild?.id) &&
        typeof channel.send === "function" &&
        typeof channel.awaitMessages === "function");
}
exports.isTextChannelInstance = isTextChannelInstance;
function isMessageInstance(message) {
    // Simple check for using distube normally
    return (!!message &&
        isSnowflake(message.id) &&
        isSnowflake(message.guild?.id) &&
        isTextChannelInstance(message.channel) &&
        isMemberInstance(message.member) &&
        isSnowflake(message.author?.id) &&
        message.member.id === message.author.id &&
        message.guild.id === message.channel.guild.id);
}
exports.isMessageInstance = isMessageInstance;
function isSupportedVoiceChannel(channel) {
    return (!!channel &&
        typeof channel.joinable === "boolean" &&
        isSnowflake(channel.id) &&
        isSnowflake(channel.guild?.id) &&
        typeof channel.full === "boolean" &&
        [
            // Djs v12
            "voice",
            "stage",
            // Djs v13
            "GUILD_VOICE",
            "GUILD_STAGE_VOICE",
        ].includes(channel.type));
}
exports.isSupportedVoiceChannel = isSupportedVoiceChannel;
function isGuildInstance(guild) {
    return !!guild && isSnowflake(guild.id) && typeof guild.fetchAuditLogs === "function";
}
exports.isGuildInstance = isGuildInstance;
function resolveGuildId(resolvable) {
    let guildId;
    if (typeof resolvable === "string") {
        guildId = resolvable;
    }
    else if (isObject(resolvable)) {
        if (resolvable instanceof _1.Queue || resolvable instanceof _1.DisTubeVoice)
            guildId = resolvable.id;
        else if ("guild" in resolvable && isGuildInstance(resolvable.guild))
            guildId = resolvable.guild.id;
        else if ("id" in resolvable && isGuildInstance(resolvable))
            guildId = resolvable.id;
    }
    if (!isSnowflake(guildId))
        throw new _1.DisTubeError("INVALID_TYPE", "GuildIdResolvable", resolvable);
    return guildId;
}
exports.resolveGuildId = resolveGuildId;
function resolveGuildID(resolvable) {
    process.emitWarning("resolveGuildID() is deprecated, use resolveGuildId() instead.", "DeprecationWarning");
    return resolveGuildId(resolvable);
}
exports.resolveGuildID = resolveGuildID;
function isClientInstance(client) {
    return !!client && typeof client.login === "function";
}
exports.isClientInstance = isClientInstance;
function checkInvalidKey(target, source, sourceName) {
    if (!isObject(target))
        throw new _1.DisTubeError("INVALID_TYPE", "object", target, sourceName);
    const sourceKeys = Array.isArray(source) ? source : Object.keys(source);
    const invalidKey = Object.keys(target).find(key => !sourceKeys.includes(key));
    if (invalidKey)
        throw new _1.DisTubeError("INVALID_KEY", sourceName, invalidKey);
}
exports.checkInvalidKey = checkInvalidKey;
async function waitEvent(target, status, maxTime) {
    let cleanup = () => undefined;
    try {
        await new Promise((resolve, reject) => {
            const timeout = setTimeout(() => reject(new Error(`Didn't trigger ${status} within ${maxTime}ms`)), maxTime);
            target.once(status, resolve);
            target.once("error", reject);
            cleanup = () => {
                clearTimeout(timeout);
                target.off(status, resolve);
                target.off("error", reject);
            };
        });
        return target;
    }
    finally {
        cleanup();
    }
}
async function entersState(target, status, maxTime) {
    if (target.state.status === status)
        return target;
    return waitEvent(target, status, maxTime);
}
exports.entersState = entersState;
function isObject(obj) {
    return typeof obj === "object" && obj !== null && !Array.isArray(obj);
}
exports.isObject = isObject;
function isRecord(obj) {
    return isObject(obj);
}
exports.isRecord = isRecord;
