// @distube/yt-dlp/dist/env.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.YTDLP_PATH = exports.YTDLP_FILENAME = exports.YTDLP_DIR = exports.YTDLP_IS_WINDOWS = exports.YTDLP_URL = exports.YTDLP_DISABLE_DOWNLOAD = void 0;
const path_1 = __importDefault(require("path"));
const e = (key) => process.env[key];
exports.YTDLP_DISABLE_DOWNLOAD = !!e("YTDLP_DISABLE_DOWNLOAD");
exports.YTDLP_URL = e("YTDLP_URL") || "https://api.github.com/repos/yt-dlp/yt-dlp/releases?per_page=1";
exports.YTDLP_IS_WINDOWS = e("YTDLP_IS_WINDOWS") || process.platform === "win32";
exports.YTDLP_DIR = e("YTDLP_DIR") || path_1.default.join(__dirname, "..", "bin");
exports.YTDLP_FILENAME = e("YTDLP_FILENAME") || `yt-dlp${exports.YTDLP_IS_WINDOWS ? ".exe" : ""}`;
exports.YTDLP_PATH = path_1.default.join(exports.YTDLP_DIR, exports.YTDLP_FILENAME);
exports.default = { YTDLP_DIR: exports.YTDLP_DIR, YTDLP_FILENAME: exports.YTDLP_FILENAME, YTDLP_URL: exports.YTDLP_URL, YTDLP_PATH: exports.YTDLP_PATH, YTDLP_IS_WINDOWS: exports.YTDLP_IS_WINDOWS, YTDLP_DISABLE_DOWNLOAD: exports.YTDLP_DISABLE_DOWNLOAD };
